storm-core/src/jvm/org/apache/storm/daemon/supervisor/Slot.java

```diff
@@ -0,0 +1,766 @@
+            return new DynamicState(this.state, this.newAssignment,
+                    this.container, this.currentAssignment,
+                    this.pendingLocalization, this.startTime,
+                    pendingDownload, this.profileActions,
+                    this.pendingStopProfileActions);
+        }
+        
+        public DynamicState withState(final MachineState state) {
+            long newStartTime = this.startTime;
+            if (state == MachineState.KILL ||
+                    state == MachineState.KILL_AND_RELAUNCH ||
+                    state == MachineState.WAITING_FOR_WORKER_START) {
+                newStartTime = Time.currentTimeMillis();
+            }
+            return new DynamicState(state, this.newAssignment,
+                    this.container, this.currentAssignment,
+                    this.pendingLocalization, newStartTime,
+                    this.pendingDownload, this.profileActions,
+                    this.pendingStopProfileActions);
+        }
+
+        public DynamicState withCurrentAssignment(final Container container, final LocalAssignment currentAssignment) {
+            return new DynamicState(this.state, this.newAssignment,
+                    container, currentAssignment,
+                    this.pendingLocalization, this.startTime,
+                    this.pendingDownload, this.profileActions,
+                    this.pendingStopProfileActions);
+        }
+
+        public DynamicState withProfileActions(Set<TopoProfileAction> profileActions, Set<TopoProfileAction> pendingStopProfileActions) {
+            return new DynamicState(this.state, this.newAssignment,
+                    this.container, this.currentAssignment,
+                    this.pendingLocalization, this.startTime,
+                    this.pendingDownload, profileActions,
+                    pendingStopProfileActions);
+        }
+    };
+    
+    static class TopoProfileAction {
+        public final String topoId;
+        public final ProfileRequest request;
+        
+        public TopoProfileAction(String topoId, ProfileRequest request) {
+            this.topoId = topoId;
+            this.request = request;
+        }
+        
+        @Override
+        public int hashCode() {
+            return (37 * topoId.hashCode()) + request.hashCode(); 
+        }
+        
+        @Override
+        public boolean equals(Object other) {
+            if (!(other instanceof TopoProfileAction)) {
+                return false;
+            }
+            TopoProfileAction o = (TopoProfileAction) other;
+            return topoId.equals(o.topoId) && request.equals(o.request);
+        }
+        
+        @Override
+        public String toString() {
+            return "{ "+topoId + ": " + request + " }";
+        }
+    }
+    
+    static boolean equivalent(LocalAssignment a, LocalAssignment b) {
+        if (a == null && b == null) {
+            return true;
+        } if (a != null && b != null) {
```

Comment:
It looks like this used to be an `else if` but can now be outdented and move to its own line.